pandora-client-web/src/graphics/graphicsCharacter.tsx

```diff
@@ -99,6 +107,35 @@ function GraphicsCharacterWithManagerImpl({
 
 	const assetPreferenceIsVisible = useAssetPreferenceVisibilityCheck();
 
+	const [characterBlinking, setCharacterBlinking] = useState(false);
+	const [blinkSeed, newBlinkSeed] = useReducer(() => nanoid(), nanoid());
+	const blinkRandom = useMemo(() => new PseudoRandom(blinkSeed), [blinkSeed]);
+
+	useEffect(() => {
+		if (!useBlinking) {
+			return;
+		}
+
+		const blinkInterval = blinkRandom.between(MIN_BLINK_INTERVAL, MAX_BLINK_INTERVAL);
+
+		let timeoutId: number | undefined;
+		const intervalId = setInterval(() => {
+			clearTimeout(timeoutId);
+			setCharacterBlinking(true);
+
+			timeoutId = setTimeout(() => {
+				setCharacterBlinking(false);
+
+				newBlinkSeed();
+			}, BLINK_LENGTH);
+		}, blinkInterval);
+
+		return () => {
+			clearInterval(intervalId);
+			clearTimeout(timeoutId);
+		};
+	}, [setCharacterBlinking, blinkRandom, newBlinkSeed, useBlinking]);
+
```

Comment:
This works, but there are a few small problems that can be improved upon:

1: `setInterval` can be replaced by `setTimeout` too. When the interval triggers, then you start a timeout for resetting blink (which is through the current values expected to trigger sooner than the outer interval). This eyes-open timeout will then call `newBlinkSeed`, which will (through cascade) regenerate `blinkRandom`, which then triggers update of this `useEffect`, clearing the interval.
In short the interval never actually triggers more than once (if everything goes well).

2: Completely regenerating the random value generator each blink is technically equivalent to just generating a fully random number each blink.
Instead I would suggest generating pseudo-random generator only once, for example using:
```ts
const blinkRandom = useMemo(() => new PseudoRandom(nanoid()), []);
```
And then having the `useEffect` do another `setTimeout` whenever the eyes get opened (getting rid of the `useReducer` altogether). That way the `useEffect`'s dependencies never change, and it isn't cleaned up and re-run each blink.
Technically this could be achieved through, for example, creating a function inside the `useEffect` and then triggering it whenever it is appropriate. Rough outline I'm thinking about:
```ts
useEffect(() => {
    let timeoutId: number | undefined;
    const run = () => {
        const blinkInterval = ...
        timeoutId = setTimeout(() => {
            // Blink here
            timeoutId = setTimeout(() => {
                // Un-blink here
                run(); // loop
            }, ...);
        }, blinkInterval);
    };
    run(); // Trigger first loop
    // Cleanup handler
    return () => {
        // Cancel `timeoutId` if set
    };
}, [...]);
```

Reply:
> `setInterval` can be replaced by `setTimeout` too
I had noticed this while looking through the commits, but evidently forgot to actually change it... Fixed now.

> 2
A very fair point I had not considered. Initially, it had been achieved through a looped function call, before moving it inside the effect. Had forgotten about inline functions. Logic has been changed accordingly, and assuming I understood correctly, the PRNG should now only be created once.